[AOTInductor] Align behavior between CPU and GPU

muchulee8 opened the pull request:
Summary:
(1) Make sure CPU and GPU doesn't have different implementation and behavior when calling from the same path and API. Only difference between CPU and GPU after this PR should ONLY be the running hardware.
(2) This PR fixes the issue of memory access when it==constants_map.end()
(3) This PR resolves T179437596

Test Plan: buck2 run mode/dev sigmoid/inference/test:e2e_test_cpu

Differential Revision: D68540744

desertfire commented on torch/csrc/inductor/aoti_runtime/model_container.h:
```diff
@@ -228,6 +226,72 @@ class AOTInductorModelContainer {
+  }
+
+  // We directly take ownership from AtenTensorHandle if constants are moved.
+  void update_constant_buffer(
```
Let's use a different name to distinguish from the later one.

Can you elaborate more on what this function is doing?  I am confused why we need aoti_torch_clone when you are talking about transferring ownership. Maybe you just need aoti_torch_assign_tensors?

muchulee8 replied:
This is doing same thing from the later one, so I prefer using the same name.
The difference between this and the later one is, this one passes in a rvalue ref, so we make the assumption that user no longer possess the tensor ownership, and we wrap it with RAII ourself. For the later one, we do a copy, so users can still do whatever they want with the tensor.

The `aoti_torch_clone` is actually a special scenario, that is some tensor that is not user-facing, (check `_should_skip_update`,) so there's no ownership w.r.t. the constant map here, those are constants being created during tracing etc.

desertfire replied:
So I guess the next function is only called through the AOTInductorModelContainerUpdateConstantBuffer interface now? My question is are you going to expose this function through another API? If yes, you need a different API. If not, naming doesn't matter much here.

muchulee8 replied:
No. We will not expose this to the user explicitly. (This is currently only used by the internal const-fold update)

muchulee8 commented on torch/csrc/inductor/aoti_runtime/model.h:
```diff
@@ -42,12 +42,26 @@ extern uint8_t _binary_constants_bin_start[];
+// Compute required blob size with 64-alignment if on GPU.
+#define AOTI_CONST_ALIGNMENT 64
+#else
+#define AOTI_CONST_ALIGNMENT 1
```
I changed it to 64. But the logic is slightly different over here, #127610 is building into the const_o, this is a allocation on the runtime memory. In theory, we could set it both to 1 over here if data_size is already being written in the codegen (since it's already 64-aligned.) I added some comment here, but still kept it separate (for CPU it's aligned and for better perf. GPU it will just break if it's not 64.)

muchulee8 commented on torch/csrc/inductor/aoti_runtime/model.h:
```diff
@@ -260,26 +274,23 @@ class AOTInductorModelBase {
-      constant_blob_ = RAII_gpuMalloc(blob_size);
+    constant_blob_ = RAII_gpuMalloc(blob_size);
+#else
+    constant_blob_ = RAII_cpuMalloc(blob_size);
```
Yes and no. The way we implement CPU is different from GPU now (hence this diff.) 
What's happening between CPU and GPU is that for GPU we always copy the data from user, while CPU do not copy data from users and fully rely on the RAII stealing mechanism. This causes a lot of down stream call because the whole memory management and when to free is different. 
So the yes part is, we will increase memory *when calling constant update*, no in that the memory for runtime is the same.

muchulee8 commented:
@pytorchbot merge

muchulee8 committed:
[AOTInductor] Align behavior between CPU and GPU (#145459)

Summary:

(1) Make sure CPU and GPU doesn't have different implementation and behavior when calling from the same path and API. Only difference between CPU and GPU after this PR should ONLY be the running hardware.
(2) This PR fixes the issue of memory access when it==constants_map.end()
(3) This PR resolves T179437596

Test Plan:
```
buck2 run mode/dev sigmoid/inference/test:e2e_test_cpu
```

Comparable for CPU memory check:
With change: 914544948_1
Without change: 914534146_1

```
914544948_1
I0206 00:24:50.725411 3424702 PtInferenceModelBench.cpp:290] HELLO WORLD BEFORE benchmark current memory usage: 1209 MB
I0206 00:24:50.725486 3424702 PtInferenceModelBench.cpp:191] Start benchmarking with threadCount = 16
W0206 00:24:50.744007 3425325 QuantizedLinear.cpp:383] Warning: fbgemm_pack_gemm_matrix_fp16 is deprecated and will be removed in a future PyTorch release. (function operator())
W0206 00:24:50.857731 3425327 QuantizedLinear.cpp:418] Warning: fbgemm_linear_fp16_weight_fp32_activation is deprecated and will be removed in a future PyTorch release. (function operator())
I0206 00:27:01.246569 3424702 PtInferenceModelBench.cpp:177] PyTorch bench finished. Average inference time (ms) =  : 2.78375
I0206 00:27:01.246610 3424702 PtInferenceModelBench.cpp:179]  number of threads = 16, total iters = 2400, total warmup iters = 208
I0206 00:27:01.246618 3424702 PtInferenceModelBench.cpp:294] HELLO WORLD AFTER benchmark current memory usage: 559 MB

914534146_1
I0206 00:28:57.804713 3495625 PtInferenceModelBench.cpp:290] HELLO WORLD BEFORE benchmark current memory usage: 1231 MB
I0206 00:28:57.804780 3495625 PtInferenceModelBench.cpp:191] Start benchmarking with threadCount = 16
W0206 00:28:57.835098 3496885 QuantizedLinear.cpp:383] Warning: fbgemm_pack_gemm_matrix_fp16 is deprecated and will be removed in a future PyTorch release. (function operator())
W0206 00:28:57.966532 3496885 QuantizedLinear.cpp:418] Warning: fbgemm_linear_fp16_weight_fp32_activation is deprecated and will be removed in a future PyTorch release. (function operator())
I0206 00:31:08.360823 3495625 PtInferenceModelBench.cpp:177] PyTorch bench finished. Average inference time (ms) =  : 2.41358
I0206 00:31:08.360870 3495625 PtInferenceModelBench.cpp:179]  number of threads = 16, total iters = 2400, total warmup iters = 208
I0206 00:31:08.360879 3495625 PtInferenceModelBench.cpp:294] HELLO WORLD AFTER benchmark current memory usage: 583 MB
```

Reviewed By: hl475

Differential Revision: D68540744

muchulee8 commented:
@pytorchbot merge -f "Merged internally, test never finish for ROCM, and related test in fbcode, can't merge from github, deadlock."
